lib/winapi/filemanip.c

```diff
@@ -235,3 +236,106 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
         return TRUE;
     }
 }
+
+BOOL GetDiskFreeSpaceExA (LPCSTR lpDirectoryName, PULARGE_INTEGER lpFreeBytesAvailableToCaller, PULARGE_INTEGER lpTotalNumberOfBytes, PULARGE_INTEGER lpTotalNumberOfFreeBytes)
+{
+    NTSTATUS status;
+    HANDLE handle;
+    ANSI_STRING path;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_FS_SIZE_INFORMATION fsSizeInfo;
+
+    assert(lpDirectoryName);
+
+    RtlInitAnsiString(&path, lpDirectoryName);
+    InitializeObjectAttributes(&objectAttributes, &path, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(&handle, FILE_LIST_DIRECTORY | SYNCHRONIZE, &objectAttributes, &ioStatusBlock, FILE_SHARE_READ | FILE_SHARE_WRITE, FILE_SYNCHRONOUS_IO_NONALERT | FILE_DIRECTORY_FILE | FILE_OPEN_FOR_FREE_SPACE_QUERY);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        if (GetLastError() == ERROR_FILE_NOT_FOUND) {
+            SetLastError(ERROR_PATH_NOT_FOUND);
+        }
+        return FALSE;
+    }
+
+    status = NtQueryVolumeInformationFile(handle, &ioStatusBlock, &fsSizeInfo, sizeof(fsSizeInfo), FileFsSizeInformation);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    status = NtClose(handle);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    ULONGLONG bytesPerAllocUnit = fsSizeInfo.BytesPerSector * fsSizeInfo.SectorsPerAllocationUnit;
+    ULONGLONG totalBytes = bytesPerAllocUnit * fsSizeInfo.TotalAllocationUnits.QuadPart;
+    ULONGLONG freeBytes = bytesPerAllocUnit * fsSizeInfo.AvailableAllocationUnits.QuadPart;
+
+    if (lpFreeBytesAvailableToCaller) {
+        lpFreeBytesAvailableToCaller->QuadPart = freeBytes;
+    }
+    if (lpTotalNumberOfBytes) {
+        lpTotalNumberOfBytes->QuadPart = totalBytes;
+    }
+    if (lpTotalNumberOfFreeBytes) {
+        lpTotalNumberOfFreeBytes->QuadPart = freeBytes;
+    }
+
+    return TRUE;
+}
+
+BOOL GetDiskFreeSpaceA (LPCSTR lpRootPathName, LPDWORD lpSectorsPerCluster, LPDWORD lpBytesPerSector, LPDWORD lpNumberOfFreeClusters, LPDWORD lpTotalNumberOfClusters)
+{
+    NTSTATUS status;
+    HANDLE handle;
+    ANSI_STRING path;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_FS_SIZE_INFORMATION fsSizeInfo;
+
+    assert(lpRootPathName);
```

Comment:
"If this parameter is NULL, the function uses the root of the current disk."
https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getdiskfreespacea

Same applies to `GetDiskFreeSpaceEx`.

Reply:
True, but we don't support the concept of current disks / current directories yet.

Reply:
Makes sense.
I personally wouldn't want the MS docs behavior anyways tbh.

Reply:
We probably do want to support it in the future, so we can set the current directory to `D:\` and then we can probably use relative paths again?

If we don't follow MS docs we'll usually document it. However, in this case we can make an exception (as there's no way to set the current folder so it's obviously problematic). Also the `assert` explicitly handles it.
I guess ideally we'd add the sentence from the MSDN to that `assert` so if people trigger it in a naive port they'll immediately know what's going on.

*(I had actually noticed this in my initial review, too but was fine with this)*